### DISCUSSION TITLE
After a module is compiled, is it possible to retrieve the optimized prompt?

### DISCUSSION THREAD

**Discussion by maximepeabody** [stanfordnlp/dspy#519] [General]
How can I output the actual prompt that the module compiled/optimized? I'm using LLMs in typescript so I can't use the module itself.

**Comment by detaos** [comment]
You can always use `lm.inspect_history(n=1)` after running the compiled program to see the prompt itself.

**Reply by RLinnae** [reply]
Can you use the compiled prompt as a starting point for future optimization runs?

Ie can you run the SignatureOptimizer.  Then use the optimized signature in. BootstrapFewShot?

Or can you run a second iteration of the SignatureOptimizer with slight variation, using the original output as the starting point?

Ok r do we always start from scratch when compiling?

**Reply by detaos** [reply]
You can definitely do a second optimization run starting from the 1st optimized prompt.  You can start from wherever you want 😄

**Reply by jingnanzhou** [reply]
can you  provide sample code how to reload the compiled status?

**Reply by lebsral** [reply]
You can save the optimized prompt.  And then load it later.
https://dspy-docs.vercel.app/docs/building-blocks/optimizers#saving-and-loading-optimizer-output

This works on all optimizers  (except COPRO?)

**Comment by kinthaiofficial** [comment]
Retrieving the optimized prompt from a compiled DSPy module is the right instinct — it's the artifact you actually care about, and being able to inspect/version/deploy it independently of the compilation framework is important for production systems.

For agent systems specifically, compiled DSPy modules raise an interesting question: what's the "prompt" when the module includes tool definitions, delegation logic, and memory retrieval steps? The optimized prompt is no longer just a text string — it's a structured program that includes control flow.

A few things we've found useful for this:

**Snapshot the full program state** — not just the prompt text but the full compiled representation: few-shot examples selected, which demonstrations were kept, the reasoning patterns the optimizer converged on. Store this as a versioned artifact.

**Separate optimization from deployment** — optimize offline, deploy the fixed compiled program. Don't re-optimize in production on every call (expensive and non-deterministic). The compiled program is the deployment artifact, like a trained model checkpoint.

**Version compiled programs against model versions** — a compiled program optimized for Claude Sonnet may perform differently on Claude Opus. Track which model version the compilation was run against.

For agent orchestration specifically, the compiled DSPy module (retrieved prompt + examples) can serve as the "skill" definition that an agent contributes to the network — other agents can invoke it without knowing the optimization process: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons has some patterns around skill composition.

Are you asking about retrieval for debugging/inspection, or for deploying to production without the full DSPy runtime?
